third-party/cmake/BuildMsgpack.cmake

```diff
@@ -1,24 +1,65 @@
-     -DMSGPACK_ENABLE_CXX=OFF
-     -DMSGPACK_BUILD_TESTS=OFF
-     -DCMAKE_INSTALL_PREFIX=${DEPS_INSTALL_DIR}
-     -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
-     "-DCMAKE_C_FLAGS:STRING=${CMAKE_C_COMPILER_ARG1} -fPIC"
-  BUILD_COMMAND ${CMAKE_COMMAND} --build .
-  INSTALL_COMMAND ${CMAKE_COMMAND}
-    -DREMOVE_FILE_GLOB=${DEPS_INSTALL_DIR}/lib/${CMAKE_SHARED_LIBRARY_PREFIX}msgpack*${CMAKE_SHARED_LIBRARY_SUFFIX}*
-    -P ${CMAKE_CURRENT_SOURCE_DIR}/cmake/InstallMsgpack.cmake)
+
+# BuildMsgpack: set MSGPACK_CONFIGURE_COMMAND, MSGPACK_BUILD_COMMAND
+#   and MSGPACK_INSTALL_COMMAND before calling this function
+function(BuildMsgpack)
+  ExternalProject_Add(msgpack
+    PREFIX ${DEPS_BUILD_DIR}
+    URL ${MSGPACK_URL}
+    DOWNLOAD_DIR ${DEPS_DOWNLOAD_DIR}/msgpack
+    DOWNLOAD_COMMAND ${CMAKE_COMMAND}
+      -DPREFIX=${DEPS_BUILD_DIR}
+      -DDOWNLOAD_DIR=${DEPS_DOWNLOAD_DIR}/msgpack
+      -DURL=${MSGPACK_URL}
+      -DEXPECTED_SHA256=${MSGPACK_SHA256}
+      -DTARGET=msgpack
+      -P ${CMAKE_CURRENT_SOURCE_DIR}/cmake/DownloadAndExtractFile.cmake
+    CONFIGURE_COMMAND "${MSGPACK_CONFIGURE_COMMAND}"
+    BUILD_COMMAND "${MSGPACK_BUILD_COMMAND}"
+    INSTALL_COMMAND ${MSGPACK_INSTALL_COMMAND})
+endfunction()
+
+set(MSGPACK_CONFIGURE_COMMAND ${CMAKE_COMMAND} ${DEPS_BUILD_DIR}/src/msgpack
+       -DMSGPACK_ENABLE_CXX=OFF
+       -DMSGPACK_BUILD_TESTS=OFF
+       -DCMAKE_INSTALL_PREFIX=${DEPS_INSTALL_DIR}
+       -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
+       "-DCMAKE_C_FLAGS:STRING=${CMAKE_C_COMPILER_ARG1} -fPIC"
+       )
+set(MSGPACK_BUILD_COMMAND ${CMAKE_COMMAND} --build .)
+set(MSGPACK_INSTALL_COMMAND ${CMAKE_COMMAND}
+      -DREMOVE_FILE_GLOB=${DEPS_INSTALL_DIR}/lib/${CMAKE_SHARED_LIBRARY_PREFIX}msgpack*${CMAKE_SHARED_LIBRARY_SUFFIX}*
+      -P ${CMAKE_CURRENT_SOURCE_DIR}/cmake/InstallMsgpack.cmake
+       )
+
+if (MINGW AND CMAKE_CROSSCOMPILING)
+  get_filename_component(TOOLCHAIN ${CMAKE_TOOLCHAIN_FILE} REALPATH)
+  set(MSGPACK_CONFIGURE_COMMAND cmake ${DEPS_BUILD_DIR}/src/msgpack
+    -DMSGPACK_ENABLE_CXX=OFF
+    -DMSGPACK_BUILD_TESTS=OFF
+    -DCMAKE_INSTALL_PREFIX=${DEPS_INSTALL_DIR}
+    # Pass toolchain
+    -DCMAKE_TOOLCHAIN_FILE=${TOOLCHAIN}
+    # Hack to avoid -rdynamic in Mingw
+    -DCMAKE_SHARED_LIBRARY_LINK_C_FLAGS="")
+elseif (MSVC)
+    # Same as UNIX without fPIC
+    set(MSGPACK_CONFIGURE_COMMAND cmake ${DEPS_BUILD_DIR}/src/msgpack
+           -DMSGPACK_ENABLE_CXX=OFF
+           -DMSGPACK_BUILD_TESTS=OFF
+           -DCMAKE_INSTALL_PREFIX=${DEPS_INSTALL_DIR}
+           -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
+           "-DCMAKE_C_FLAGS:STRING=${CMAKE_C_COMPILER_ARG1}"
+	   # Make sure we use the same generator, otherwise we may
+	   # accidentaly end up using different MSVC runtimes
+	   -DCMAKE_GENERATOR=${CMAKE_GENERATOR}
+           # Use static runtime
+           -DCMAKE_C_FLAGS_DEBUG="-MTd"
+           -DCMAKE_C_FLAGS_RELEASE="-MT")
+endif()
+
+BuildMsgpack()
 list(APPEND THIRD_PARTY_DEPS msgpack)
 
+
+
```

Comment:
We should try to avoid the blank lines at the end of the files.  Git likes to complain about them in some situations. :-)